zkevm-circuits/src/tx_circuit/sign_verify.rs

```diff
@@ -504,60 +512,97 @@ impl<F: Field> SignVerifyChip<F> {
 
         let pk_le = pk_bytes_le(&sign_data.pk);
         let pk_be = pk_bytes_swap_endianness(&pk_le);
-        let pk_hash = (!padding)
-            .then(|| keccak256(&pk_be))
-            .unwrap_or_default()
-            .map(|byte| Value::known(F::from(byte as u64)));
-        let pk_hash_hi = pk_hash[..12].to_vec();
+        let mut pk_hash = (!padding).then(|| keccak256(&pk_be)).unwrap_or_default();
+        pk_hash.reverse();
+
+        let powers_of_256 = iter::successors(Some(F::ONE), |coeff| Some(F::from(256) * coeff))
+            .take(20)
+            .collect_vec();
+
         // Ref. spec SignVerifyChip 2. Verify that the first 20 bytes of the
         // pub_key_hash equal the address
-        let (address, pk_hash_lo) = {
-            let powers_of_256 = iter::successors(Some(F::ONE), |coeff| Some(F::from(256) * coeff))
-                .take(20)
-                .collect_vec();
-            let terms = pk_hash[12..]
-                .iter()
-                .zip(powers_of_256.into_iter().rev())
-                .map(|(byte, coeff)| maingate::Term::Unassigned(*byte, coeff))
-                .collect_vec();
-            let (address, pk_hash_lo) = main_gate.decompose(ctx, &terms, F::ZERO, |_, _| Ok(()))?;
+        let (address_cells, pk_hash_cells) = {
+            // Diagram of byte decomposition of little-endian pk_hash, and how address is built
+            // from it:
+            //
+            // byte 0             15 16           20 21   32
+            //      [ address_lo   ] [ address_hi  ] [     ]
+            //      [ pk_hash_lo   ] [ pk_hash_hi          ]
+
+            let pk_hash_lo_bytes = &pk_hash[..16];
+            let pk_hash_hi_bytes = &pk_hash[16..];
+            let pk_hash_lo = from_bytes::value::<F>(pk_hash_lo_bytes);
+            let pk_hash_hi = from_bytes::value::<F>(pk_hash_hi_bytes);
+            // Assign all bytes of pk_hash to cells which are range constrained to be 8 bits.  Then
+            // constrain the lower 16 cell bytes to build the lo cell, and the higher 16 bytes to
+            // build the hi cell.
+            let (pk_hash_cell_lo, pk_hash_lo_cell_bytes) =
+                range_chip.decompose(ctx, Value::known(pk_hash_lo), 8, 128)?;
+            let (pk_hash_cell_hi, pk_hash_hi_cell_bytes) =
+                range_chip.decompose(ctx, Value::known(pk_hash_hi), 8, 128)?;
+
+            // Take the 20 lowest assigned byte cells of pk_hash and constrain them to build
+            // address. From the lower 16 build the lo cell, and from the higher 4 build the hi
+            // cell.
+            let (address_cell_lo, _) = main_gate.decompose(
+                ctx,
+                &pk_hash_lo_cell_bytes
+                    .iter()
+                    .zip(&powers_of_256)
```

Comment:
here `zip_eq` is better than `zip`

Reply:
fixed in [24c67ee](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1418/commits/24c67eea3617aaa417a24fd5e1a0f21895fe06f7)